Use std::span more with SharedBuffer

cdumez opened the pull request:
#### 20f975626f24cd2a36d74011d4c6f70a3895feb2
<pre>
Use std::span more with SharedBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=271534">https://bugs.webkit.org/show_bug.cgi?id=271534</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::mutableSpan):
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.cpp:
(WebCore::ClipboardItemBindingsDataSource::ClipboardItemTypeLoader::didFinishLoading):
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::put):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::resolve):
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::consumeChunk):
* Source/WebCore/Modules/highlight/AppHighlightRangeData.cpp:
(WebCore::AppHighlightRangeData::toSharedBuffer const):
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::appendBuffer):
(WebCore::SourceBuffer::appendBufferInternal):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCDataChannelHandler.cpp:
(WebCore::LibWebRTCDataChannelHandler::OnMessage):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::create):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create): Deleted.
(WebCore::FragmentedSharedBuffer::append): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.h:
* Source/WebCore/platform/encryptedmedia/CDMProxy.h:
(WebCore::KeyHandle::idAsSharedBuffer const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::extractKeyidsFromCencInitData):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::CDMPrivateFairPlayStreaming::mptsKeyIDs):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData):
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandle::readDataAsync):
(WebCore::BlobResourceHandle::didRead):
(WebCore::BlobResourceHandle::consumeData):
(WebCore::BlobResourceHandle::notifyReceiveData):
* Source/WebCore/platform/network/BlobResourceHandle.h:
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::append):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::Record::didReceiveResponseBodyChunk):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::processResponse):

Canonical link: <a href="https://commits.webkit.org/276672@main">https://commits.webkit.org/276672@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbbc54c402026e885f60a33f6efa7ee3c0ae32a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45321 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47848 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47985 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41329 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21838 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45899 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/21498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18268 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40178 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3368 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/38541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41590 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49696 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/44790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/20304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16831 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/45362 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39255 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43011 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51949 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6308 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21299 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -800,6 +800,7 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     size_t capacity() const { return Base::capacity(); }
     bool isEmpty() const { return !size(); }
     std::span<const T> span() const { return { data(), size() }; }
+    std::span<T> mutableSpan() { return { data(), size() }; }
```
If we instead did this just as a non-const overload on `span()`, do we run into compiler errors?

cdumez replied:
Yes, for example:
```
./runtime/IntlCollator.cpp:501:25: error: no matching function for call to 'charactersAreAllASCII'
                    if (charactersAreAllASCII(buffer.span())) {
```

darinadler commented on Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
```diff
@@ -248,8 +248,8 @@ RefPtr<SharedBuffer> CDMPrivateFairPlayStreaming::sanitizeMpts(const SharedBuffe
-        Ref keyBuffer = SharedBuffer::create(keyID.characters(), keyID.length());
+        const auto keyID = "TransportStreamIdentifier"_span;
+        Ref keyBuffer = SharedBuffer::create(keyID);
         return Vector { 1, WTFMove(keyBuffer) };
```
I would write it more like this:
```suggestion
        return Vector { 1, SharedBuffer::create("TransportStreamIdentifier"_span) };
```
And we might be able to put two sets of braces instead of the `1` to get an `initializer_list`?

darinadler commented on Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
```diff
@@ -298,8 +298,8 @@ CDMInstanceSessionThunder::CDMInstanceSessionThunder(CDMInstanceThunder& instanc
         GST_ERROR("Got 'error' OCDM notification: %s", message);
-        callOnMainThread([session = WeakPtr { static_cast<CDMInstanceSessionThunder*>(userData) }, buffer = WebCore::SharedBuffer::create(message,
-            strlen(message))]() mutable {
+        callOnMainThread([session = WeakPtr { static_cast<CDMInstanceSessionThunder*>(userData) }, buffer = WebCore::SharedBuffer::create(std::span { message,
```
Seems like "span of a C-style string using `strlen`" is a recurring pattern. Should we add a helper for that?

cdumez replied:
Planning to add these, let me know you'd prefer something different:
```
inline std::span<const LChar> span8(const char* string)
{
    return { reinterpret_cast<const LChar*>(string), string ? strlen(string) : 0 };
}

inline std::span<const char> span(const char* string)
{
    return { string, string ? strlen(string) : 0 };
}
```

darinadler replied:
I think we might want the name it slightly longer to make it clear that it does the `strlen` operation. The other span functions are just getting a "view" on something, so they are O(1), but this is O(string length). It’s also interesting that with `CString` we found we also needed the one that includes the terminator byte.

darinadler replied:
Trying to think of a name but `nullTerminatedStringSpan` is too long. And `stringSpan` isn’t obviously talking about C strings.

cdumez replied:
What what's it's worth, I looked at the code base to try and adopt my new span() / span8() as much as possible. One of the call sites I couldn't port because it wanted a span that included the null terminator. All others were fine with either span() or span8(). The short name didn't really bother me, even though it doesn't call strlen(). Seems clear to me that it has to call strlen if you're not provided a length.

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -342,9 +342,9 @@ int BlobResourceHandle::readSync(uint8_t* buf, int length)
         int bytesRead = 0;
         if (item.type() == BlobDataItem::Type::Data)
-            bytesRead = readDataSync(item, buf + offset, remaining);
+            bytesRead = readDataSync(item, buffer.subspan(offset, remaining));
```
```suggestion
            bytesRead = readDataSync(item, buffer.subspan(offset));
```

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -342,9 +342,9 @@ int BlobResourceHandle::readSync(uint8_t* buf, int length)
+            bytesRead = readDataSync(item, buffer.subspan(offset, remaining));
         else if (item.type() == BlobDataItem::Type::File)
-            bytesRead = readFileSync(item, buf + offset, remaining);
+            bytesRead = readFileSync(item, buffer.subspan(offset, remaining));
```
```suggestion
            bytesRead = readFileSync(item, buffer.subspan(offset));
```

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -358,28 +358,28 @@ int BlobResourceHandle::readSync(uint8_t* buf, int length)
-        bytesToRead = static_cast<int>(m_totalRemainingSize);
-    memcpy(buf, item.data()->data() + item.offset() + m_currentItemReadSize, bytesToRead);
+    size_t remaining = item.length() - m_currentItemReadSize;
+    size_t bytesToRead = buffer.size() > remaining ? remaining : buffer.size();
```
Maybe we can do something like:
```
size_t bytesToRead = std::min({ static_cast<uint64_t>(m_totalRemainingSize), remaining, static_cast<uint64_t>(buffer.size()) }));
```
We can use local variables to do the type casts if we don’t like the length of the line. To use `std::min` we have to cast them all to the same type.

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -358,28 +358,28 @@ int BlobResourceHandle::readSync(uint8_t* buf, int length)
-    if (bytesToRead > m_totalRemainingSize)
-        bytesToRead = static_cast<int>(m_totalRemainingSize);
-    memcpy(buf, item.data()->data() + item.offset() + m_currentItemReadSize, bytesToRead);
+    size_t remaining = item.length() - m_currentItemReadSize;
```
Is `size_t` big enough here? Maybe we need `uint64_t`? After doing the `std::min` can chop down to `size_t`, but beforehand that might not be OK. Unless the thing we are reading from is in memory and so we are guaranteed it is smaller than `size_t`. The old code didn’t chop before comparing `length` with `remaining`.

cdumez replied:
That's a good point. I'll make sure to not chop down to size_t too early.

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.h:
```diff
@@ -79,20 +79,20 @@ class BlobResourceHandle final : public FileStreamClient, public ResourceHandle
     void getSizeForNext();
     std::optional<Error> seek();
-    void consumeData(const uint8_t* data, int bytesRead);
+    void consumeData(std::span<const uint8_t> data);
```
I don’t think we need the argument name here.

darinadler commented on Source/WebCore/workers/ScriptBuffer.cpp:
```diff
@@ -86,7 +86,7 @@ void ScriptBuffer::append(const String& string)
     if (string.isEmpty())
         return;
     auto result = string.tryGetUTF8([&](std::span<const char> span) -> bool {
```
Later we should change the type passed by `tryGetUTF8` to use `char8_t` instead of `char`. But we’ll still need `spanReinterpretCast` here, so we’re good.

darinadler commented on Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
```diff
@@ -462,11 +462,11 @@ void NetworkDataTaskBlob::download()
     ASSERT(isDownload());
-    int bytesWritten = FileSystem::writeToFile(m_downloadFile, data, bytesRead);
-    if (bytesWritten != bytesRead) {
+    int bytesWritten = FileSystem::writeToFile(m_downloadFile, data.data(), data.size());
```
What guarantees that `data.size()` isn’t chopped when passing it as an `int`? I think we need to look out for newly-introduced cases where large numbers might be chopped and treated as small numbers. This is one of them, but we should look around for others.

darinadler commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -313,13 +313,13 @@ std::optional<CacheStorageRecord> CacheStorageDiskStore::readRecordFromFileData(
             return std::nullopt;
 
-        auto bodyData = std::span(buffer.data() + bodyOffset, bodySize);
+        std::span bodyData = buffer.subspan(bodyOffset, bodySize);
```
Could have left this `auto` instead of writing `std::span`.

darinadler commented on Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
```diff
@@ -168,7 +168,7 @@ void Clipboard::readBuffer(const char* format, CompletionHandler<void(Ref<WebCor
         int contentsLength;
         const auto* contents = gtk_selection_data_get_data_with_length(selection, &contentsLength);
-        data->completionHandler(WebCore::SharedBuffer::create(contents, contentsLength > 0 ? static_cast<size_t>(contentsLength) : 0));
+        data->completionHandler(WebCore::SharedBuffer::create(std::span { contents, contentsLength > 0 ? static_cast<size_t>(contentsLength) : 0 }));
```
```suggestion
        data->completionHandler(WebCore::SharedBuffer::create(std::span { contents, static_cast<size_t>(std::max(0, contentsLength)) }));
```

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
```diff
@@ -80,20 +81,20 @@ TEST_F(FragmentedSharedBufferTest, appendBufferCreatedWithContentsOfExistingFile
-    builder.append(testData2, strlen(testData2));
+    const auto testData0 = "Hello"_span;
+    const auto testData1 = "World"_span;
+    const auto testData2 = "Goodbye"_span;
```
`constexpr`? Or don’t use local variables?

cdumez replied:
`_span` doesn't currently work with constexpr. It relies on bitwise_cast() which calls memcpy() which is not allowed in constexpr context.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
```diff
@@ -413,25 +414,25 @@ TEST_F(SharedBufferChunkReaderTest, includeSeparator)
     for (size_t i = 0; i < 256; ++i)
         data[i] = i;
-    auto sharedBuffer = SharedBuffer::create(data, 256);
+    auto sharedBuffer = SharedBuffer::create(std::span<const uint8_t> { data });
```
Is the `std::span` needed? Just to add the `const`?

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
```diff
@@ -413,25 +414,25 @@ TEST_F(SharedBufferChunkReaderTest, includeSeparator)
     check(sharedBuffer);
 
-    SharedBufferBuilder builder(std::in_place, data, 256);
+    SharedBufferBuilder builder(std::in_place, std::span<const uint8_t> { data });
```
Is the `std::span` needed? Just to add the `const`?

darinadler reviewed (approved):
I’m concerned about some mixing of integer sizes, but otherwise looks great.

cdumez committed:
Use std::span more with SharedBuffer
https://bugs.webkit.org/show_bug.cgi?id=271534

Reviewed by Darin Adler.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::mutableSpan):
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.cpp:
(WebCore::ClipboardItemBindingsDataSource::ClipboardItemTypeLoader::didFinishLoading):
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::put):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::resolve):
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::consumeChunk):
* Source/WebCore/Modules/highlight/AppHighlightRangeData.cpp:
(WebCore::AppHighlightRangeData::toSharedBuffer const):
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::appendBuffer):
(WebCore::SourceBuffer::appendBufferInternal):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCDataChannelHandler.cpp:
(WebCore::LibWebRTCDataChannelHandler::OnMessage):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::create):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create): Deleted.
(WebCore::FragmentedSharedBuffer::append): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.h:
* Source/WebCore/platform/encryptedmedia/CDMProxy.h:
(WebCore::KeyHandle::idAsSharedBuffer const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::extractKeyidsFromCencInitData):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::CDMPrivateFairPlayStreaming::mptsKeyIDs):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData):
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandle::readDataAsync):
(WebCore::BlobResourceHandle::didRead):
(WebCore::BlobResourceHandle::consumeData):
(WebCore::BlobResourceHandle::notifyReceiveData):
* Source/WebCore/platform/network/BlobResourceHandle.h:
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::append):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::Record::didReceiveResponseBodyChunk):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::processResponse):

Canonical link: https://commits.webkit.org/276672@main
